File: frontend/src/api/modules/power.ts
/**
 * 供配电管理 API
 */
import request from '@/utils/request'

// ===== 类型定义 =====

export interface PowerOverviewSummary {
  ups_total: number
  ups_online: number
  ups_offline: number
  ups_alarm: number
  battery_total: number
  battery_avg_soh: number
  battery_lowest_soc: number
  cabinet_total: number
  pdu_total: number
  total_load_kw: number
  avg_load_rate: number
}

export interface UPSDeviceInfo {
  id: number
  device_id: number
  ups_type: string
  rated_capacity: number
  rated_voltage: number
  phase_count: number
  battery_group_count: number
  bypass_enabled: boolean
  description: string | null
  created_at: string
  updated_at: string
  // 关联的 Device 信息（由后端 join 返回）
  device_code?: string
  device_name?: string
  area_code?: string
  status?: string
}

export interface UPSDeviceCreate {
  device_id: number
  ups_type: string
  rated_capacity: number
  rated_voltage: number
  phase_count: number
  battery_group_count?: number
  bypass_enabled?: boolean
  description?: string | null
}

export interface UPSDeviceUpdate {
  ups_type?: string
  rated_capacity?: number
  rated_voltage?: number
  phase_count?: number
  battery_group_count?: number
  bypass_enabled?: boolean
  description?: string | null
}

export interface BatteryGroupInfo {
  id: number
  ups_device_id: number
  group_name: string
  battery_type: string
  rated_capacity: number
  rated_voltage: number
  cell_count: number
  install_date: string | null
  description: string | null
  created_at: string
  updated_at: string
}

export interface BatteryGroupCreate {
  ups_device_id: number
  group_name: string
  battery_type: string
  rated_capacity: number
  rated_voltage: number
  cell_count: number
  install_date?: string | null
  description?: string | null
}

export interface BatteryGroupUpdate {
  group_name?: string
  battery_type?: string
  rated_capacity?: number
  rated_voltage?: number
  cell_count?: number
  install_date?: string | null
  description?: string | null
}

export interface PointRealtimeValue {
  point_code: string
  point_name: string
  value: number | null
  unit: string
  status: string
}

// ===== API 调用 =====

/** 获取供配电总览 */
export function getPowerOverview() {
  return request.get<PowerOverviewSummary>('/v1/power/overview')
}

/** 获取 UPS 列表 */
export function getUPSList(params?: { page?: number; page_size?: number; status?: string }) {
  return request.get<any>('/v1/power/ups', { params })
}

/** 获取 UPS 详情（含实时点位数据） */
export function getUPSDetail(id: number) {
  return request.get<any>(`/v1/power/ups/${id}`)
}

/** 创建 UPS 设备 */
export function createUPS(data: UPSDeviceCreate) {
  return request.post<any>('/v1/power/ups', data)
}

/** 更新 UPS 设备 */
export function updateUPS(id: number, data: UPSDeviceUpdate) {
  return request.put<any>(`/v1/power/ups/${id}`, data)
}

/** 删除 UPS 设备 */
export function deleteUPS(id: number) {
  return request.delete<any>(`/v1/power/ups/${id}`)
}

/** 获取电池组列表 */
export function getBatteryList(params?: { page?: number; page_size?: number }) {
  return request.get<any>('/v1/power/batteries', { params })
}

/** 获取电池组详情 */
export function getBatteryDetail(id: number) {
  return request.get<any>(`/v1/power/batteries/${id}`)
}

/** 创建电池组 */
export function createBatteryGroup(data: BatteryGroupCreate) {
  return request.post<any>('/v1/power/batteries', data)
}

/** 更新电池组 */
export function updateBatteryGroup(id: number, data: BatteryGroupUpdate) {
  return request.put<any>(`/v1/power/batteries/${id}`, data)
}

/** 删除电池组 */
export function deleteBatteryGroup(id: number) {
  return request.delete<any>(`/v1/power/batteries/${id}`)
}

/** 获取配电柜列表 */
export function getCabinetList(params?: { page?: number; page_size?: number }) {
  return request.get<any>('/v1/power/cabinets', { params })
}

/** 获取配电柜支路详情 */
export function getCabinetBranches(deviceId: number) {
  return request.get<any>(`/v1/power/cabinets/${deviceId}/branches`)
}

/** 获取 PDU 列表 */
export function getPDUList(params?: { page?: number; page_size?: number }) {
  return request.get<any>('/v1/power/pdus', { params })
}

File: backend/app/schemas/power.py
"""
供配电管理 Schema
"""

from typing import Optional
from datetime import datetime, date
from pydantic import BaseModel, ConfigDict


# ========== UPS设备 ==========


class UPSDeviceCreate(BaseModel):
    """创建UPS设备"""

    device_id: int
    ups_type: Optional[str] = "standalone"
    rated_capacity: Optional[float] = None
    rated_voltage: Optional[float] = None
    phase_count: Optional[int] = 3
    battery_group_count: Optional[int] = 1
    bypass_enabled: Optional[bool] = True
    description: Optional[str] = None


class UPSDeviceUpdate(BaseModel):
    """更新UPS设备"""

    ups_type: Optional[str] = None
    rated_capacity: Optional[float] = None
    rated_voltage: Optional[float] = None
    phase_count: Optional[int] = None
    battery_group_count: Optional[int] = None
    bypass_enabled: Optional[bool] = None
    description: Optional[str] = None


class UPSDeviceInfo(BaseModel):
    """UPS设备信息"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    device_id: int
    ups_type: Optional[str] = "standalone"
    rated_capacity: Optional[float] = None
    rated_voltage: Optional[float] = None
    phase_count: int = 3
    battery_group_count: int = 1
    bypass_enabled: bool = True
    description: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ========== 电池组 ==========


class BatteryGroupCreate(BaseModel):
    """创建电池组"""

    ups_device_id: int
    group_name: str
    battery_type: Optional[str] = "lead_acid"
    rated_capacity: Optional[float] = None
    rated_voltage: Optional[float] = None
    cell_count: Optional[int] = None
    install_date: Optional[date] = None
    description: Optional[str] = None


class BatteryGroupUpdate(BaseModel):
    """更新电池组"""

    group_name: Optional[str] = None
    battery_type: Optional[str] = None
    rated_capacity: Optional[float] = None
    rated_voltage: Optional[float] = None
    cell_count: Optional[int] = None
    install_date: Optional[date] = None
    description: Optional[str] = None


class BatteryGroupInfo(BaseModel):
    """电池组信息"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    ups_device_id: Optional[int] = None
    group_name: Optional[str] = None
    battery_type: Optional[str] = "lead_acid"
    rated_capacity: Optional[float] = None
    rated_voltage: Optional[float] = None
    cell_count: Optional[int] = None
    install_date: Optional[date] = None
    description: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ========== 供配电总览 ==========


class PowerOverviewSummary(BaseModel):
    """供配电总览统计"""

    ups_total: int = 0
    ups_online: int = 0
    ups_offline: int = 0
    ups_alarm: int = 0
    battery_total: int = 0
    battery_avg_soh: float = 0.0
    battery_lowest_soc: float = 0.0
    cabinet_total: int = 0
    pdu_total: int = 0
    total_load_kw: float = 0.0
    avg_load_rate: float = 0.0
